File: checks/src/main/java/com/example/lint/checks/TimerTaskLeakDetector.java
package com.example.lint.checks;

import com.android.tools.lint.client.api.UElementHandler;
import com.android.tools.lint.detector.api.Category;
import com.android.tools.lint.detector.api.Detector;
import com.android.tools.lint.detector.api.Implementation;
import com.android.tools.lint.detector.api.Issue;
import com.android.tools.lint.detector.api.JavaContext;
import com.android.tools.lint.detector.api.Scope;
import com.android.tools.lint.detector.api.Severity;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.jetbrains.uast.UElement;

import java.util.EnumSet;
import java.util.List;


public class TimerTaskLeakDetector extends Detector implements Detector.UastScanner {

    private static final String ID = "TimerTaskLeak";
    private static final String DESCRIPTION = "TimerTask should be canceled";
    private static final String EXPLANATION = "We should cancel TimerTask when activity is destroyed";
    private static final Category CATEGORY = Category.PERFORMANCE;
    private static final int PRIORITY = 4;
    private static final Severity SEVERITY = Severity.WARNING;

    public static final Issue ISSUE = Issue.create(
            ID,
            DESCRIPTION,
            EXPLANATION,
            CATEGORY,
            PRIORITY,
            SEVERITY,
            new Implementation(TimerTaskLeakDetector.class, EnumSet.of(Scope.JAVA_FILE))
    );

    @Nullable
    @Override
    public UElementHandler createUastHandler(@NotNull JavaContext context) {
        return super.createUastHandler(context);
    }

    @Nullable
    @Override
    public List<Class<? extends UElement>> getApplicableUastTypes() {
        return super.getApplicableUastTypes();
    }
}

File: checks/src/main/java/com/example/lint/checks/TimerTaskLeakDetector.kt
package com.example.lint.checks

import com.android.tools.lint.client.api.UElementHandler
import com.android.tools.lint.detector.api.*
import com.android.tools.lint.detector.api.Category.Companion.PERFORMANCE
import org.jetbrains.uast.UElement
import java.util.*

class TimerTaskLeakDetector : Detector(), Detector.UastScanner {
    override fun createUastHandler(context: JavaContext): UElementHandler? {
        return super.createUastHandler(context)
    }

    override fun getApplicableUastTypes(): List<Class<out UElement?>>? {
        return super.getApplicableUastTypes()
    }

    companion object {
        private const val ID = "TimerTaskLeak"
        private const val DESCRIPTION = "TimerTask should be canceled"
        private const val EXPLANATION = "We should cancel TimerTask when activity is destroyed"
        private val CATEGORY: Category = PERFORMANCE
        private const val PRIORITY = 4
        private val SEVERITY = Severity.WARNING

        @JvmField
        val ISSUE = Issue.create(
            ID,
            DESCRIPTION,
            EXPLANATION,
            CATEGORY,
            PRIORITY,
            SEVERITY,
            Implementation(TimerTaskLeakDetector::class.java, EnumSet.of(Scope.JAVA_FILE))
        )
    }
}